packages/context/src/context.ts

```diff
@@ -34,16 +43,90 @@ export class Context {
   protected _parent?: Context;
 
   /**
-   * Create a new context
+   * A list of registered context observers
+   */
+  protected readonly observers: Set<ContextObserver> = new Set();
+
+  /**
+   * Internal counter for pending events which observers have not processed yet
+   */
+  private pendingEvents = 0;
+
+  /**
+   * Create a new context. For example,
+   * ```ts
+   * // Create a new root context, let the framework to create a unique name
+   * const rootCtx = new Context();
+   *
+   * // Create a new child context inheriting bindings from `rootCtx`
+   * const childCtx = new Context(rootCtx);
+   *
+   * // Create another root context called "application"
+   * const appCtx = new Context('application');
+   *
+   * // Create a new child context called "request" and inheriting bindings
+   * // from `appCtx`
+   * const reqCtx = new Context(appCtx, 'request');
+   * ```
    * @param _parent The optional parent context
+   * @param name Name of the context, if not provided, a `uuid` will be
+   * generated as the name
    */
   constructor(_parent?: Context | string, name?: string) {
+    super();
     if (typeof _parent === 'string') {
       name = _parent;
       _parent = undefined;
     }
     this._parent = _parent;
     this.name = name || uuidv1();
+
+    this.setupEventHandlers();
+  }
+
+  /**
+   * Set up an internal listener to notify registered observers asynchronously
+   * upon `bind` and `unbind` events
+   */
+  private setupEventHandlers() {
+    // Ideally p-event should allow multiple event types in an iterator
+    this.observeEvent('bind');
+    this.observeEvent('unbind');
+  }
+
+  /**
+   * Listen on context events and notify observers
+   * @param eventType Context event type
+   */
+  private async observeEvent(eventType: ContextEventType) {
+    this.on(eventType, () => {
+      // Track pending events
+      this.pendingEvents++;
+    });
+    // Create an async iterator from the given event type
+    const bindings: AsyncIterable<Readonly<Binding<unknown>>> = pEvent.iterator(
+      this,
+      eventType,
+    );
+    for await (const binding of bindings) {
+      try {
+        await this.notifyObservers(eventType, binding);
+        this.pendingEvents--;
+        this.emit('idle');
+      } catch (err) {
+        this.pendingEvents--;
+        this.emit('error', err);
```

Comment:
If no error listeners are registered with the context, the process will crash.

Reply:
I find this approach slightly problematic.

As an application developer, I can easily attach `on('error')` handler to the root (application-level) context, but there is no easy way how to attach error handler to child contexts like per-request contexts created for each incoming request. As a result, we end up in a setup where failure of certain observers will trigger our error handler while failure of other observers will not.

I think this is an edge case that will be interesting to very few people, so it's not worth to make our code too complex. At the same time, I don't want our users to be surprised about inconsistency in LB behavior.

I am proposing one of the following changes:

(1) When a new child context is created, it should propagate error events to its parent. Conceptually:

```ts
class Context {
  constructor(parent?: Context) {
    if (parent) this.on('error', (...args) => parent.emit('error', ...args));
  }
}
```

This way it's easy for application developers to setup error handler that will be called for errors from all context observers.

(2) Don't use `error` event at all, print errors to console or crash the process. This may be too drastic, the first option described above is probably a better one.